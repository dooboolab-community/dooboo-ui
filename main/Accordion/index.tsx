import {Animated, StyleProp, ViewStyle} from 'react-native';
import React, {FC, useRef} from 'react';

import AccordionItem from './AccordionItem';
import styled from '@emotion/native';

const Container = styled.View`
  flex-direction: column;
  align-items: center;
`;

export interface Datum {
  title: string;
  bodies: string[];
}

export type Styles = {
  titleContainer?: StyleProp<ViewStyle>;
  bodyContainer?: StyleProp<ViewStyle>;
};

interface Props {
  style?: ViewStyle;
  styles?: Styles;
  data: Datum[];
  shouldAnimate?: boolean;
  collapseOnStart?: boolean;
  animDuration?: number;
  activeOpacity?: number;
  toggleElement?: React.ReactElement;
  renderTitle?: (item: string) => React.ReactElement;
  renderBody?: (item: string) => React.ReactElement;
}

const Accordion: FC<Props> = (props) => {
  const {
    style,
    styles,
    data,
    collapseOnStart,
    shouldAnimate,
    animDuration,
    activeOpacity,
    toggleElement,
    renderTitle,
    renderBody,
  } = props;

  const dropDownAnimValueList = useRef(
    data.map(() => new Animated.Value(0)),
  ).current;

  return (
    <Container style={style}>
      {data.map((datum, titleKey) => {
        return (
          <AccordionItem
            key={titleKey}
            testID={`${titleKey}`}
            styles={styles}
            datum={datum}
            collapseOnStart={collapseOnStart}
            shouldAnimate={shouldAnimate}
            animDuration={animDuration}
            activeOpacity={activeOpacity}
            dropDownAnimValueList={dropDownAnimValueList[titleKey]}
            sumOfPrecedingTranslateY={dropDownAnimValueList
              .filter((item, idx) => idx < titleKey)
              .map((value) => ({translateY: value}))}
            toggleElement={toggleElement}
            renderTitle={renderTitle}
            renderBody={renderBody}
          />
        );
      })}
    </Container>
  );
};

export {Accordion};
