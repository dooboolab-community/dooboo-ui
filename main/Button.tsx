import {
  ActivityIndicator,
  LayoutRectangle,
  Platform,
  TouchableOpacity,
} from 'react-native';
import {DoobooTheme, light, withTheme} from './theme';
import React, {useRef, useState} from 'react';
import type {
  StyleProp,
  TextProps,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';

import type {FC} from 'react';
import styled from '@emotion/native';
import {useHover} from 'react-native-web-hooks';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

const Container = styled.View`
  align-self: stretch;
  padding: 8px 12px;
  background-color: ${({theme}) => theme.primary || light.primary};

  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

const Text = styled.Text`
  font-size: 14px;
  color: white;
`;

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
}

const StyledButton: FC<ButtonProps & {theme: DoobooTheme}> = ({
  testID,
  theme,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = theme.disabled,
  leftElement,
  rightElement,
  activeOpacity = 0.7,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
}) => {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  const compositeStyles: Styles = {
    disabledButton: {
      backgroundColor: theme.primary,
      borderColor: theme.primary,
    },
    disabledText: {
      color: theme.primary,
    },
    hovered: {
      borderColor: theme.primary,
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <Container
          testID="loading-view"
          style={[
            compositeStyles.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </Container>
      ) : (
        <Container
          testID="button-view"
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}>
          {leftElement}
          <Text
            style={[
              compositeStyles.text,
              disabled && compositeStyles.disabledText,
            ]}
            {...textProps}>
            {text}
          </Text>
          {rightElement}
        </Container>
      )}
    </TouchableOpacity>
  );
};

StyledButton.defaultProps = {theme: light};

export const Button = withTheme(StyledButton);
