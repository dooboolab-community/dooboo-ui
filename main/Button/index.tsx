import {
  ActivityIndicator,
  LayoutRectangle,
  Platform,
  TouchableOpacity,
} from 'react-native';
import {
  ButtonText,
  ButtonType,
  ButtonWrapper,
} from '../Styled/StyledComponents';
import {DoobooTheme, light, withTheme} from '../theme';
import React, {useRef, useState} from 'react';
import type {
  StyleProp,
  TextProps,
  TextStyle,
  TouchableOpacityProps,
  ViewStyle,
} from 'react-native';
import styled, {css} from '@emotion/native';

import type {FC} from 'react';
import {useHover} from 'react-native-web-hooks';

type Styles = {
  container?: StyleProp<ViewStyle>;
  text?: StyleProp<TextStyle>;
  disabledButton?: StyleProp<ViewStyle>;
  disabledText?: StyleProp<TextStyle>;
  hovered?: StyleProp<ViewStyle>;
};

type ButtonSize = 'small' | 'medium' | 'large';

const ButtonContainer = styled(ButtonWrapper)<{
  type: ButtonType;
  size?: ButtonSize;
  outlined?: boolean;
  disabled?: boolean;
}>`
  align-self: stretch;
  padding: ${({size}) =>
    size === 'large'
      ? '14px 24px'
      : size === 'small'
      ? '4px 12px'
      : '10px 20px'};
  border-radius: ${({size}) => (size === 'large' ? '24px' : '20px')};

  flex-direction: row;
  align-items: center;
  justify-content: center;
`;

export interface ButtonProps {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  outlined?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
  type?: ButtonType;
  size?: ButtonSize;
}

const ButtonComponent: FC<ButtonProps & {theme: DoobooTheme}> = ({
  testID,
  theme,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = theme.text,
  leftElement,
  rightElement,
  activeOpacity = 0.6,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
  type = 'primary',
  size,
  outlined = false,
}) => {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  const compositeStyles: Styles = {
    disabledButton: css`
      background-color: ${!outlined ? theme.disabled : theme.background};
      border-color: ${theme.disabled};
    `,
    disabledText: {
      color: !outlined ? theme.background : theme.textDisabled,
    },
    hovered: {
      shadowColor: 'black',
      shadowOffset: {
        width: 0,
        height: 4,
      },
      shadowOpacity: 0.24,
      shadowRadius: 16,
      elevation: 10,
    },
    ...styles,
  };

  return (
    <TouchableOpacity
      testID={testID}
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled || loading}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <ButtonContainer
          testID="loading-view"
          disabled
          style={[
            compositeStyles.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          type={type}
          size={size}
          outlined={outlined}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </ButtonContainer>
      ) : (
        <ButtonContainer
          testID="button-container"
          style={[
            compositeStyles.container,
            hovered && !disabled && compositeStyles.hovered,
            disabled && compositeStyles.disabledButton,
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}
          type={type}
          size={size}
          disabled={disabled}
          outlined={outlined}>
          {leftElement}
          <ButtonText
            outlined={outlined}
            type={type}
            style={[
              compositeStyles.text,
              disabled && compositeStyles.disabledText,
            ]}
            {...textProps}>
            {text}
          </ButtonText>
          {rightElement}
        </ButtonContainer>
      )}
    </TouchableOpacity>
  );
};

ButtonComponent.defaultProps = {theme: light};

export const Button = withTheme(ButtonComponent);
