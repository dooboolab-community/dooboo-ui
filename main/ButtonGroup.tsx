import {DoobooTheme, light, useTheme, withTheme} from './theme';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

interface Styles {
  container: StyleProp<ViewStyle>;
  button: ViewStyle;
  selectedButton: ViewStyle;
  text: StyleProp<TextStyle>;
  selectedText: StyleProp<TextStyle>;
}

interface Props<T> {
  testID?: string;
  theme: DoobooTheme;
  borderRadius?: number;
  borderWidth?: number;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  data: T[];
  color?: string;
  onPress?: (i: number) => void;
  selected?: number;
}

function StyledButtonGroup<T>(props: Props<T>): React.ReactElement {
  const {theme} = useTheme();

  const {
    borderRadius = 0,
    selected = 0,
    borderWidth = 1,
    color = theme.text,
    testID,
    style,
    data,
    onPress,
    styles,
  } = props;

  return (
    <View
      testID={testID}
      style={StyleSheet.flatten([
        {borderColor: color},
        styles?.container,
        style,
      ])}>
      {data.map((text, i) => {
        return (
          <TouchableOpacity
            key={i}
            testID={`CHILD_${i}`}
            activeOpacity={0.85}
            style={{flex: 1}}
            onPress={(): void => {
              if (onPress) onPress(i);
            }}>
            <View
              style={StyleSheet.flatten([
                selected === i
                  ? {...styles?.selectedButton, backgroundColor: color}
                  : {...styles?.button, borderColor: color},
                i === 0
                  ? {
                      borderLeftWidth: borderWidth,
                      borderTopWidth: borderWidth,
                      borderBottomWidth: borderWidth,
                      borderTopLeftRadius: borderRadius,
                      borderBottomLeftRadius: borderRadius,
                    }
                  : i === data.length - 1
                  ? {
                      borderRightWidth: borderWidth,
                      borderLeftWidth: borderWidth,
                      borderTopWidth: borderWidth,
                      borderBottomWidth: borderWidth,
                      borderBottomRightRadius: borderRadius,
                      borderTopRightRadius: borderRadius,
                    }
                  : {
                      borderLeftWidth: borderWidth,
                      borderTopWidth: borderWidth,
                      borderBottomWidth: borderWidth,
                    },
                {borderColor: color},
              ])}>
              <Text
                style={
                  selected === i
                    ? [styles?.selectedText, {color: theme.textContrast}]
                    : [styles?.text, {color: theme.text}]
                }>
                {text}
              </Text>
            </View>
          </TouchableOpacity>
        );
      })}
    </View>
  );
}

StyledButtonGroup.defaultProps = {
  theme: light,
  styles: {
    container: {
      backgroundColor: 'transparent',
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      alignSelf: 'stretch',
      minHeight: 40,
      marginTop: 24,
    },
    button: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    selectedButton: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    text: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
    selectedText: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
  },
  data: ['option 1', 'option 2'],
};

export const ButtonGroup = withTheme<any>(StyledButtonGroup);
