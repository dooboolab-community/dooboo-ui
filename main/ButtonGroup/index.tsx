import React from 'react';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

import {useTheme} from '@dooboo-ui/theme';

interface Styles {
  container: StyleProp<ViewStyle>;
  button: ViewStyle;
  selectedButton: ViewStyle;
  text: StyleProp<TextStyle>;
  selectedText: StyleProp<TextStyle>;
}

interface Props<T> {
  testID?: string;
  borderRadius?: number;
  borderWidth?: number;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  data: T[];
  color?: string;
  onPress?: (i: number) => void;
  selectedIndex?: number;
}

export function ButtonGroup<T>(props: Props<T>): React.ReactElement {
  const {theme} = useTheme();

  const {
    borderRadius = 0,
    selectedIndex = 0,
    borderWidth = 1,
    color = theme.text,
    testID,
    style,
    data,
    onPress,
    styles,
  } = props;

  const borderWidthAndRadius = (index: number): object => {
    const fullWidthAndRadius = {
      borderLeftWidth: borderWidth,
      borderRightWidth: borderWidth,
      borderTopWidth: borderWidth,
      borderBottomWidth: borderWidth,

      borderTopLeftRadius: borderRadius,
      borderBottomLeftRadius: borderRadius,
      borderTopRightRadius: borderRadius,
      borderBottomRightRadius: borderRadius,
    };

    const isFirst = index === 0;
    const isLast = index === data.length - 1;

    const borderForFirstElement = {
      ...fullWidthAndRadius,
      borderTopRightRadius: undefined,
      borderBottomRightRadius: undefined,
    };

    const borderForLastElement = {
      ...fullWidthAndRadius,
      borderTopLeftRadius: undefined,
      borderBottomLeftRadius: undefined,
    };

    const borderForMiddleElement = {
      borderRightWidth: borderWidth,
      borderTopWidth: borderWidth,
      borderBottomWidth: borderWidth,
    };

    if (data.length === 1) return fullWidthAndRadius;

    if (isFirst) return borderForFirstElement;

    if (isLast) {
      if (data.length === 2)
        return {
          ...borderForLastElement,
          borderLeftWidth: undefined,
        };

      return borderForLastElement;
    }

    return borderForMiddleElement;
  };

  return (
    <View
      testID={testID}
      style={StyleSheet.flatten([
        {borderColor: color},
        styles?.container,
        style,
      ])}
    >
      {data.map((text, i) => {
        return (
          <TouchableOpacity
            key={i}
            activeOpacity={0.85}
            style={{flex: 1}}
            onPress={(): void => {
              if (onPress) onPress(i);
            }}
          >
            <View
              testID={`CHILD_${i}`}
              style={StyleSheet.flatten([
                selectedIndex === i
                  ? {...styles?.selectedButton, backgroundColor: color}
                  : {...styles?.button, borderColor: color},
                borderWidthAndRadius(i),
                {borderColor: color},
              ])}
            >
              <Text
                style={
                  selectedIndex === i
                    ? [styles?.selectedText, {color: theme.textContrast}]
                    : [styles?.text, {color: theme.text}]
                }
              >
                {text}
              </Text>
            </View>
          </TouchableOpacity>
        );
      })}
    </View>
  );
}

ButtonGroup.defaultProps = {
  styles: {
    container: {
      backgroundColor: 'transparent',
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      alignSelf: 'stretch',
      minHeight: 40,
      marginTop: 24,
    },
    button: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    selectedButton: {
      alignSelf: 'stretch',
      minHeight: 40,

      justifyContent: 'center',
      alignItems: 'center',
    },
    text: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
    selectedText: {
      fontSize: 14,
      textAlign: 'center',
      alignSelf: 'center',
    },
  },
  data: ['option 1', 'option 2'],
};
