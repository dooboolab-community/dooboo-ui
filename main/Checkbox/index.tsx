import {Animated, Platform, StyleProp, View, ViewStyle} from 'react-native';
import {
  CheckboxWrapper,
  CheckboxWrapperOutlined,
} from '../Styled/StyledComponents';
import React, {FC, useEffect, useRef} from 'react';

import {Icon} from '../Icon';
import styled from '@emotion/native';
import {DoobooTheme, useTheme} from '..';

type Styles = {
  checkbox?: StyleProp<ViewStyle>;
};

export type CheckboxType =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'danger'
  | 'warning'
  | 'info';

export interface CheckboxProps {
  onPress?: () => void;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  type?: CheckboxType;
  disabled?: boolean;
  checked?: boolean;
  rightElement?: React.ReactElement;
  leftElement?: React.ReactElement;
}

const Container = styled.TouchableOpacity`
  flex-direction: row;
  align-items: center;
`;

const StyledCheckboxOutlined = styled(CheckboxWrapperOutlined)<{
  checked?: boolean;
  disabled?: boolean;
  type?: CheckboxType;
}>`
  width: 20px;
  height: 20px;
  border-width: 1px;
  margin: 0 6px;

  justify-content: center;
  align-items: center;
`;

const StyledCheckbox = styled(CheckboxWrapper)<{
  checked?: boolean;
  disabled?: boolean;
  type?: CheckboxType;
}>`
  width: 20px;
  height: 20px;
  margin: 0 6px;

  justify-content: center;
  align-items: center;
`;

const StyledCheck = styled(Icon)<{theme: DoobooTheme; checked?: boolean}>`
  color: ${({theme, checked}) => (checked ? theme.background : 'transparent')};
`;

export const Checkbox: FC<CheckboxProps> = ({
  style,
  styles,
  rightElement,
  leftElement,
  type = 'primary',
  disabled = false,
  checked = false,
  onPress,
}) => {
  const animatedValue = new Animated.Value(0);
  const fadeAnim = useRef(animatedValue).current;
  const scaleAnim = useRef(animatedValue).current;

  const {theme} = useTheme();

  useEffect(() => {
    Animated.sequence([
      Animated.spring(fadeAnim, {
        toValue: !checked ? 0 : 1,
        useNativeDriver: Platform.select({
          web: false,
          default: true,
        }),
      }),
      Animated.spring(scaleAnim, {
        toValue: !checked ? 0 : 1,
        useNativeDriver: Platform.select({
          web: false,
          default: true,
        }),
      }),
    ]).start();
  }, [fadeAnim, scaleAnim, checked]);

  return (
    <Container
      disabled={disabled}
      style={style}
      onPress={onPress}
      activeOpacity={0.9}
    >
      <View
        style={{
          flexDirection: 'row',
          paddingVertical: 6,
          paddingLeft: leftElement ? 6 : 0,
          paddingRight: rightElement ? 6 : 0,
        }}
      >
        {leftElement}
        <StyledCheckboxOutlined
          testID="doobooui-checkbox"
          checked={checked}
          type={type}
          disabled={disabled}
          style={styles?.checkbox}
        >
          <StyledCheckbox
            style={[
              styles?.checkbox,
              {opacity: fadeAnim, transform: [{scale: scaleAnim}]},
            ]}
            checked={checked}
            type={type}
            disabled={disabled}
          >
            <StyledCheck theme={theme} name="tick-light" checked={checked} />
          </StyledCheckbox>
        </StyledCheckboxOutlined>
        {rightElement}
      </View>
    </Container>
  );
};
