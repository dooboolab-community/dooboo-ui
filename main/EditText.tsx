import {DoobooTheme, light} from './theme';
import {Platform, Text, TextInput, View} from 'react-native';
import React, {FC, LegacyRef, useRef, useState} from 'react';
import type {
  StyleProp,
  TextInputProps,
  TextStyle,
  ViewStyle,
} from 'react-native';

import {useHover} from 'react-native-web-hooks';
import {withTheme} from '@emotion/react';

type Styles = {
  container?: StyleProp<ViewStyle>;
  hovered?: StyleProp<ViewStyle>;
  labelText?: StyleProp<TextStyle>;
  labelTextHovered?: StyleProp<TextStyle>;
  input?: StyleProp<TextStyle>;
  errorText?: StyleProp<TextStyle>;
  counter?: StyleProp<TextStyle>;
};

export type EditTextProps = {
  testID?: TextInputProps['testID'];
  inputRef?: LegacyRef<TextInput>;
  theme?: DoobooTheme;
  textInputProps?: TextInputProps;
  style?: StyleProp<ViewStyle>;
  styles?: Styles;
  labelText?: string;
  errorText?: string;
  value?: TextInputProps['value'];
  onChange?: TextInputProps['onChange'];
  onChangeText?: TextInputProps['onChangeText'];
  placeholder?: TextInputProps['placeholder'];
  placeholderColor?: TextInputProps['placeholderTextColor'];
  onFocus?: TextInputProps['onFocus'] | undefined;
  onBlur?: TextInputProps['onBlur'] | undefined;
  editable?: TextInputProps['editable'];
  autoCapitalize?: TextInputProps['autoCapitalize'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onSubmitEditing?: TextInputProps['onSubmitEditing'];

  focusColor?: string;
  hoverColor?: string;
  errorColor?: string;
  disableColor?: string;
  labelColor?: string;
  type?: 'row' | 'column' | 'boxed';
};

const Component: FC<EditTextProps & {theme: DoobooTheme}> = ({
  theme,
  testID,
  inputRef,
  textInputProps,
  style,
  styles,
  labelText = '',
  errorText = '',
  value = '',
  placeholder,
  onChange,
  onChangeText,
  onFocus,
  onBlur,
  onSubmitEditing,
  autoCapitalize = 'none',
  secureTextEntry = false,
  editable = true,
  placeholderColor = theme.placeholder,
  focusColor = theme.primary,
  hoverColor = theme.primary,
  errorColor = theme.danger,
  disableColor = theme.disabled,
  labelColor = theme.placeholder,
  type = 'column',
}) => {
  const [focused, setFocused] = useState(false);
  const ref = useRef<View>(null);
  const hovered = useHover(ref);

  const borderColor = theme.text;
  const textColor = theme.text;

  const compositeStyles: Styles =
    type === 'column'
      ? {
          container: [
            {
              alignSelf: 'stretch',
              justifyContent: 'space-between',
              borderBottomWidth: (focused && editable) || errorText ? 2 : 1,
              borderBottomColor: borderColor,

              flexDirection: 'column',
            },
            styles?.container,
          ],
          hovered: [
            {
              borderBottomWidth: 2,
              borderBottomColor: hoverColor,
            },
            styles?.hovered,
          ],
          labelText: [
            {
              paddingHorizontal: 4,
              fontSize: 14,
              color: labelColor,
            },
            styles?.labelText,
          ],
          labelTextHovered: [
            {
              paddingHorizontal: 4,
              color: hoverColor,
            },
            styles?.labelTextHovered,
          ],
          input: [
            {
              paddingHorizontal: 10,
              paddingVertical: 10,
              fontSize: 14,
              color: !editable ? disableColor : textColor,
            },
            styles?.input,
          ],
          errorText: [
            {
              marginTop: 8,
              paddingHorizontal: 10,
              fontSize: 12,
              color: errorColor,
            },
            styles?.errorText,
          ],
          counter: [
            {
              marginTop: 8,
              paddingHorizontal: 10,
              fontSize: 12,
              color: !editable ? disableColor : textColor,
            },
            styles?.counter,
          ],
        }
      : type === 'row'
      ? {
          container: [
            {
              alignSelf: 'stretch',
              flexDirection: 'row',
              alignItems: 'center',
              justifyContent: 'space-between',
              borderBottomWidth: focused || errorText ? 2 : 1,
              borderBottomColor: borderColor,
            },
            styles?.container,
          ],
          hovered: [
            {
              borderBottomColor: hoverColor,
              borderBottomWidth: 2,
            },
            styles?.hovered,
          ],
          labelText: [
            {
              fontSize: 14,
              color: labelColor,
              marginRight: 2,
            },
            styles?.labelText,
          ],
          labelTextHovered: [
            {
              color: hoverColor,
            },
            styles?.labelTextHovered,
          ],
          input: [
            {
              paddingVertical: 10,
              paddingHorizontal: 8,
              fontSize: 14,
              flex: 1,
              color: !editable ? disableColor : textColor,
            },
            styles?.input,
          ],
          errorText: [
            {
              marginTop: 4,
              fontSize: 12,
              color: errorColor,
            },
            styles?.errorText,
          ],
          counter: [
            {
              marginTop: 4,
              fontSize: 12,
              color: !editable ? disableColor : textColor,
            },
            styles?.counter,
          ],
        }
      : {
          container: [
            {
              alignSelf: 'stretch',
              justifyContent: 'space-between',
              borderWidth: focused || errorText ? 2 : 1,
              borderColor,

              flexDirection: 'column',
            },
            styles?.container,
          ],
          hovered: [
            {
              borderWidth: 1,
              borderColor: hoverColor,
            },
            styles?.hovered,
          ],
          labelText: [
            {
              paddingHorizontal: 8,
              marginTop: 6,
              fontSize: 14,
              color: labelColor,
            },
            styles?.labelText,
          ],
          labelTextHovered: [
            {
              paddingHorizontal: 4,
              color: hoverColor,
            },
            styles?.labelTextHovered,
          ],
          input: [
            {
              textAlignVertical: 'top',
              paddingHorizontal: 10,
              paddingVertical: 10,
              fontSize: 14,
              color: !editable ? disableColor : textColor,
              height: 120,
            },
            styles?.input,
          ],
          errorText: [
            {
              marginTop: 8,
              paddingHorizontal: 10,
              fontSize: 12,
              color: errorColor,
            },
            styles?.errorText,
          ],
          counter: [
            {
              marginTop: 8,
              paddingHorizontal: 10,
              fontSize: 12,
              color: !editable ? disableColor : textColor,
            },
            styles?.counter,
          ],
        };

  return (
    <View
      testID="container-id"
      ref={Platform.select({
        web: ref,
        default: undefined,
      })}
      style={[{alignSelf: 'stretch', flexDirection: 'column'}, style]}>
      <View
        style={[
          compositeStyles.container,
          editable && hovered && compositeStyles.hovered,
          {
            borderColor: !editable
              ? disableColor
              : errorText
              ? errorColor
              : hovered && !focused
              ? hoverColor
              : focused
              ? focusColor
              : borderColor,
          },
          type !== 'boxed' && {
            borderBottomColor: !editable
              ? disableColor
              : errorText
              ? errorColor
              : hovered && !focused
              ? hoverColor
              : focused
              ? focusColor
              : borderColor,
          },
        ]}>
        {labelText ? (
          <Text
            style={[
              compositeStyles.labelText,
              hovered && editable && compositeStyles.labelTextHovered,
              !editable
                ? {color: disableColor}
                : errorText
                ? {color: errorColor}
                : hovered && !focused
                ? {color: hoverColor}
                : focused
                ? {color: focusColor}
                : {},
            ]}>
            {labelText}
          </Text>
        ) : null}
        <TextInput
          testID={testID}
          ref={inputRef}
          autoCapitalize={autoCapitalize}
          secureTextEntry={secureTextEntry}
          style={[
            compositeStyles.input,
            // @ts-ignore
            Platform.OS === 'web' && {outlineWidth: 0},
          ]}
          editable={editable}
          onFocus={(e) => {
            setFocused(true);
            onFocus?.(e);
          }}
          onBlur={(e) => {
            setFocused(false);
            onBlur?.(e);
          }}
          multiline={type === 'boxed'}
          value={value}
          placeholder={placeholder}
          placeholderTextColor={placeholderColor}
          onChange={onChange}
          onChangeText={onChangeText}
          onSubmitEditing={onSubmitEditing}
          {...textInputProps}
        />
      </View>
      {editable ? (
        <View
          style={{
            flex: 1,
            flexDirection: 'row',
            justifyContent: 'space-between',
          }}>
          <Text
            style={[compositeStyles.errorText, {flex: 1, color: errorColor}]}>
            {errorText}
          </Text>
          {textInputProps?.maxLength && (
            <Text
              style={
                value.length < textInputProps.maxLength
                  ? compositeStyles.counter
                  : [compositeStyles.errorText, {color: errorColor}]
              }>{`${value.length}/${textInputProps.maxLength}`}</Text>
          )}
        </View>
      ) : null}
    </View>
  );
};

Component.defaultProps = {theme: light};

export const EditText = withTheme(Component);
