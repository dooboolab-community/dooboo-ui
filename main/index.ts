export * from './uis/Typography';
export * from './uis/RadioGroup';
export * from './modals/Snackbar';
export * from './uis/Icon';
export * from './uis/Hr';
export * from './uis/Button';
export * from './uis/ButtonGroup';
export * from './uis/IconButton';
export * from './uis/Checkbox';
export * from './uis/EditText';
export * from './uis/NetworkImage';
export * from './uis/ProgressBar';
export * from './uis/ProgressCircle';
export * from './uis/LoadingIndicator';
export * from './uis/Accordion';
export * from './uis/StatusbarBrightness';
export * from './uis/SwitchToggle';
export * from './uis/Fabs';
export * from './uis/Rating';
export * from './uis/CustomPressable';
export * from './providers';
