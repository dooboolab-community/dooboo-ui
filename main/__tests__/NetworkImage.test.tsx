import * as React from 'react';

import {createComponent, createTestProps} from '../../test/testUtils';

import {NetworkImage} from '../../main';
import {View} from 'react-native';
import renderer from 'react-test-renderer';

let props: any;
let component: React.ReactElement;

describe('[NetworkImage] render', () => {
  beforeEach(() => {
    props = createTestProps();

    component = createComponent(<NetworkImage {...props} />);
  });

  it('should render without crashing', () => {
    const rendered = renderer.create(component).toJSON();

    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});
