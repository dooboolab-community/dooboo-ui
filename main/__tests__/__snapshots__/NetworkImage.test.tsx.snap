// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[NetworkImage] render should render without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      false,
      undefined,
    ]
  }
>
  <Image
    resizeMethod="resize"
    resizeMode="cover"
    source={
      Object {
        "canInstrument": true,
        "createTransformer": [Function],
        "getCacheKey": [Function],
        "process": [Function],
      }
    }
    style={
      Array [
        Object {
          "aspectRatio": 1.4864864864864864,
          "height": "auto",
          "maxHeight": "50%",
          "width": "50%",
        },
        undefined,
      ]
    }
  />
</View>
`;
